Fix `is_subgraph_isomorphic` with parallel edges.

georgios-ts opened the pull request:
Closes #429.
The solution is to build an adjacency matrix with entries equal
to the number of edges connecting a pair of nodes instead of
0/1 values.

<!--
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I ran rustfmt locally
- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

mtreinish commented on src/isomorphism/vf2.rs:
```diff
@@ -41,6 +38,48 @@ use crate::iterators::NodeMap;
 
+/// Returns the adjacency_matrix of a graph with `(i, j)` entry equal
+/// to number of edges from node `i` to node `j`.
+fn adjacency_matrix<Ty: EdgeType>(graph: &StablePyGraph<Ty>) -> Vec<usize> {
```
I'm wondering if instead of using a matrix it would be better to fix #367 as part of this too and just query the graph object to get the edge count on demand. Not sure what the performance difference will be though, but it will at least scale better in the amount of memory required.

georgios-ts replied:
Yeah, I thought about it but in the interest of time, I decided to push this fix and in a follow up, once we also have a serious benchmark, we can check if there is a time performance difference using something like `HashMap`

mtreinish commented:
When this is combined with #418 can we drop the fixedbitset dependency from the `Cargo.toml`?

IvanIsCoding commented on src/isomorphism/vf2.rs:
```diff
@@ -41,6 +38,48 @@ use crate::iterators::NodeMap;
+/// to number of edges from node `i` to node `j`.
+fn adjacency_matrix<Ty: EdgeType>(graph: &StablePyGraph<Ty>) -> Vec<usize> {
+    let n = graph.node_bound();
+    let mut matrix = vec![0; n * n];
```
Because of #367, I would say using a sparse matrix or hashmap would be more appropriate if we stick with the adjacency matrix approach

georgios-ts commented:
This benchmark https://github.com/Qiskit/retworkx/pull/288#issuecomment-804448939 runs in `3.77 sec` with `Vec` while with `HashMap` in `2.64 sec`. So, I switched to `HashMap` in b0745b1 and we can close #367 with this PR.

I addressed also the further bugs reported https://github.com/Qiskit/retworkx/issues/429#issuecomment-910171179 in 538ba51

mtreinish commented on src/isomorphism/vf2.rs:
```diff
@@ -41,6 +38,86 @@ use crate::iterators::NodeMap;
+fn adjacency_matrix<Ty: EdgeType>(
+    graph: &StablePyGraph<Ty>,
+) -> HashMap<(NodeIndex, NodeIndex), usize> {
+    let mut matrix = HashMap::with_capacity(graph.edge_count());
```
Won't this potentially overallocate in the case of parallel edges?

georgios-ts replied:
Hmm, yeah, it might over allocate but probably by not that much. It's quite difficult to obtain a more tight bound in the number of non-parallel edges. Are you suggesting to not pre-allocate at all?

mtreinish replied:
Yeah, that's what I was getting at. But, thinking about it some more I agree it's probably not a bid deal we basically overallocate 3 usizes for each parallel edge in the graph which will probably hardly be noticeable. Especially compared to the extra allocation overhead for not pre-allocating.

mtreinish commented on src/isomorphism/vf2.rs:
```diff
@@ -41,6 +38,86 @@ use crate::iterators::NodeMap;
+    if !graph.is_directed() {
+        sorted(&mut item);
+    }
+    matrix.get(&item).cloned().unwrap_or(0) >= val
```
Do we need to actually clone here? Can't we just match on the get and use 0 if None, something like:
```suggestion
    match matrix.get(&item) {
        Some(count) => count >= val,
        None => 0 >= val, 
    } 
```

mtreinish replied:
True, the overhead should be small, it just stood out as unnecessary. Thanks for updating it.

mtreinish reviewed (commented):
Overall LGTM, nice catch and fix. A couple inline comments. Also maybe we should add another fix note about the decrease in memory usage for #367?

mtreinish commented:
Also when this is ready I think between this and #423 that's probably big enough to start to prepare a 0.10.2 release. I'll create a stable branch and start staging that after this merges.

IvanIsCoding commented on src/isomorphism/vf2.rs:
```diff
@@ -41,6 +38,86 @@ use crate::iterators::NodeMap;
+    b: NodeIndex,
+    val: usize,
+) -> bool {
+    let mut item = (a, b);
+    if !graph.is_directed() {
+        sorted(&mut item);
+    }
+    *matrix.get(&item).unwrap_or(&0) >= val
```
```suggestion
    edge_multiplicity(&graph, &matrix, a, b) >= val
```

I think the shorter way makes it more obvious

mtreinish reviewed (approved):
LGTM, thanks for fixing this.
